src/mono/mono/metadata/assembly.c

```diff
@@ -3188,6 +3204,164 @@ mono_register_bundled_satellite_assemblies (const MonoBundledSatelliteAssembly *
 	satellite_bundles = assemblies;
 }
 
+static GHashTable *bundled_resources = NULL;
+
+static int *resources_bundled_counts = NULL;
+
+//---------------------------------------------------------------------------------------
+//
+// mono_add_bundled_resource handles bundling of many types of resources to circumvent
+// needing to find or have those resources on disk. The BundledResource struct models
+// the union of information carried by all supported types of resources which are
+// enumerated in BundledResourceType.
+//
+// bundled_resources:
+// A single hash table will hold all resources being bundled with the understanding
+// that all resources being bundled are uniquely named. The bundled resource is tagged
+// with the type of resource it represents for later processing of specific types of
+// resources as done in mono_register_bundled_resources.
+//
+// resources_bundled_counts:
+// An integer array tracks the number of each resource type added to the bundle
+// to facilitate appropriate memory allocation when processing the bundle as done in
+// mono_register_bundled_resources
+//
+// Arguments:
+//  * name - Unique name of the resource
+//  * culture - Culture of satellite assemblies
+//  * data - Byte data array of the bundled resource
+//  * size - Size of byte data array of the bundled resource
+//  type - Type of bundled resource
+//
+
+void
+mono_add_bundled_resource (const char *name, const char *culture, const unsigned char *data, unsigned int size, BundledResourceType type)
+{
```

Comment:
We know all the bundled resources statically right?
And they're all already in memory?

So why do we add them one at a time, instead of getting some kind of bulk data structure that already has all the memory allocated.

Maybe we could take some pre-allocated structures and a pre-computer hash array for lookups  and avoid allocating one `BundledResource` at a time and inserting them into a hash table at startup.

Make the MSBuild task do more work ahead of time.

Reply:
From what I understand, the `EmitBundle` task will generate a source file that primarily declares a function that adds one-by-one, each resource's generated byte array data and len to be bundled
i.e. 
```
void mono_register_resources_bundle() {
  mono_add_bundled_resource (...)
  mono_add_bundled_resource (...)
  ...
  mono_add_bundled_resource (...)
  
  // finally register assemblies and satellite assemblies
  mono_register_bundled_resources ()
}
```
Each resource's generated byte array data and len symbols are linked into the final self-contained library. Does that mean that all of the data is in memory when the self contained library is loaded? So all that would need to be done is to invoke `mono_register_bundled_assemblies`, `mono_register_bundled_satellite_assemblies`, and `mono_register_symfile_for_assembly`?

Creating the hash table was primarily taking the approach done in https://github.com/dotnet/runtime/blob/b014ab87e4c25232807dc36e6f2f9138a716ab47/src/mono/mono/utils/mono-dl-wasm.c#L100-L136 and organizing in a way that `mono_register_bundled_assemblies` could be invoked for all assembly bundled resources, `mono_register_bundled_satellite_assemblies` for all satellite assembly bundled resources, and still be able to grab data for a given bundled resource on demand, like https://github.com/dotnet/runtime/blob/b014ab87e4c25232807dc36e6f2f9138a716ab47/src/native/libs/System.Native/pal_datetime.c#L70, and for grabbing the bundled runtimeconfig.bin in this PR for an adapted form of this https://github.com/dotnet/runtime/blob/b014ab87e4c25232807dc36e6f2f9138a716ab47/src/tasks/LibraryBuilder/Templates/autoinit.c#L30

> So why do we add them one at a time, instead of getting some kind of bulk data structure that already has all the memory allocated.

What would the bulk data structure look like? In one way or another, the memory to allocate `MonoBundledAssembly **` and `MonoBundledSatelliteAssembly **` would be needed, right?

> Maybe we could take some pre-allocated structures and a pre-computer hash array for lookups and avoid allocating one BundledResource at a time and inserting them into a hash table at startup.
> Make the MSBuild task do more work ahead of time

So for the MSBuild task, instead of generating a source file for each resource that contains
```
unsigned char Android_Device_Emulator_Aot_Llvm_Test_dll_339BD88C_c[] = { ... byte data ... }
unsigned int Android_Device_Emulator_Aot_Llvm_Test_dll_339BD88C_c_len = 5120;
```
, we create through the MSBulid task a hash array of resource name to their information?

The thing we'd have to still do at startup, would be registering the assembly and satellite assemblies right?

Reply:
Is each call to the `EmitBundle` task passed the complete set of resources of a certain type (ie all assemblies or all BUNDLED_BINARY, etc)?

If so, then I imagine the task could generate something like this:

```c
// suppose this is for assemblies
void mono_register_resources_bundle(void) {
  static const char *filenames[] = {
    "System.Private.CoreLib.dll",
    "System.Whatever.dll",
    NULL
  };
  static const char spc_data[] =     { /* bytes for System.Private.CoreLib.dll */ };
  static const char whatever_data[] = { /* bytes for System.Whatever.dll. */ };
  static const char  byte_data[] = { spc_data, whatever_data, NULL};
  static uint32_t lengths[] = {
    12345 /* length of System.Private.CoreLib.dll data */,
    23456 /* length of System.Whatever.dll */,
    -1
  };

  // buckets that are indexed by the name stored in buckets.
  static uint32_t num_buckets = 17;
  static uint32_t buckets[] = {
     0, /* empty buckets are filled wth zero */
     1, /* non-empty buckets are filled with 1+ the index of the corresponding filenames, and byte_data arrays */,
     0,
     2, /* if there are hash collisions, use some re-hashing scheme to go look in another bucket */
     ...,
     0,
  }

  mono_bulk_add_bundled_resources (MONO_BUNDLE_ASSEMBLY, 2/* count of elements */, filenames, byte_data, lengths, num_buckets, buckets);
}
```

Then all that `mono_bulk_add_bundled_resources` is just stash away the stuff that's passed in.

And to lookup a resource by name you hash the name and look for the bucket.  Then the bucket tells you the index into the `filenames`.  Then if the filename matches the one you want, you return the corresponding `byte_data[i]` and `lengths[i]`.  If not, you re-hash using whatever rehash scheme you decide to use.

Something like that.

Reply:
> Is hardcoding hast table worth it ? I vaguely prefer the dynamic approach. For reasons that we could have multiple modules/libraries which will bring their own set of files and could add the set by multiple calls. We could also make the generation and compilation of the registration code more incremental. Right now, it's always generated in full.

I think it might be different for different scenarios.  I think on mobile it probably doesn't make sense to do multiple separate registrations.

I think what I outlined above could do both.  For example we could make the bulk registration function that does incremental additions if you pass `0, NULL` for `hash_size` and `hash_buckets` or a pre-computed hash table for cases where we have everything.

```c
/// Bulk add resources: either incremental or all at once.  If incremental, pass 0 for hash_size and NULL for hash_buckets
/// If all at once with a pre-computed hash table, pass > 0 for hash_size and non-null hash buckets
mono_bulk_add_bundled_resources (MonoBundleResourceType type, int assm_count, const char **filenames, const uint8_t **resource_data, uint32_t *resource_lengths, int hash_size, const int *hash_buckets);
```

We could even implement this interface right now, but assert if `hash_size != 0 || hack_buckets != NULL` - ie leave room for us for future expansion to support a pre-computed hash.

Reply:
> Is each call to the EmitBundle task passed the complete set of resources of a certain type

Right now its passed all resources of any type. Maybe I can tweak the task to group the different files to generate the suggested structure.

Reply:
Is this approach primarily to enhance performance on startup?
It saves the space and time needed to hash things minus the cost of `mono_bulk_add_bundled_resources`, but is that expected to be that costly in the first place?

Reply:
> Is this approach primarily to enhance performance on startup? It saves the space and time needed to hash things minus the cost of `mono_bulk_add_bundled_resources`, but is that expected to be that costly in the first place?

Yea, I'm primarily concerned about startup. Ideally it would be great if we didn't have to allocate too much at runtime just to populate some data structures that we pretty much understand at build-time.

I don't know if it's costly, but historically the XA team has worried a lot about startup perf, and I think since this is an early bottleneck for everything it might start to matter for a self-contained library with a couple hundred assemblies.

But I don't know. we might need to measure.

Reply:
As I said, one approach we can take is to just leave ourselves room to support bulk allocation in the future.

I guess none of these APIs are truly public, so we can iterate as we learn more, but it's worth at least thinking about whether we can do better.  And maybe there's some cheap things we can tweak now.

Reply:
New logic has been added to support preallocation of `MonoBundled*Resource` structs done in the `EmitBundle` task, so the only allocation being done now is to allocate the HashTable that stores the pointers of the preallocated `MonoBundled*Resource`s